// harbors/harbors_modules/harbors.js
//NODEJS模块
//NODEJS modules
exports.fs = require('fs');
exports.net = require('net');
exports.sys = require('sys');
exports.url = require('url');
exports.path = require('path');
exports.zlib = require('zlib');
exports.http = require('http');
exports.crypto = require('crypto');
exports.cluster = require('cluster');
exports.querystring = require('querystring');

//所有模块的入口文件
//Entrance file for all modules
exports.SERVER = 'HARBORS/0.3.2';
exports.DEBUG = false;
exports.BASEDIR = exports.path.join(__dirname,'../../../');//basic dir
exports.DEBUG ?
    exports.log = function(){
        var str;
        for(var p in arguments){
            if(!str) str = arguments[p]
            else str += arguments[p];
        }
        console.log(str);
    } :
    exports.log = function(){};
exports.error = function(str){ console.error(str); };


//HARBORS模块
//HARBORS modules
exports.$mime = require('./mime');
exports.$file = require('./file');
exports.$date = require('./date');
exports.$sync = require('./sync');
exports.$utils = require('./utils');
exports.$server = require('./server');
exports.$cluster = require('./cluster');
exports.$handle = require('./handle');
exports.$module = require('./module');
exports.$fastcgi = require('./fastcgi');

//HARBORS缓存
//HARBORS cache
exports._listen = [];
exports._vhost = [];
exports._rewrite = {};
exports._config = require('../config/config');
