// harbors/lib/object.js

var _display = require('./display');
var _state = require('./state');
var _session = require('./session');
var _cookie = require('./cookie');
var _send = require('./send');
var _render = require('./render');
var _param = require('./param');
var _files = require('./files');
var _path = require('./path');

var Modular = function(config,vconfig,req,res){
    return new Modular.fn.init(config,vconfig,req,res);
};

Modular.prototype = {};

Modular.prototype.init = function(config,vconfig,req,res){
    this.config = config;
    this.vhost = vconfig;
    this.req = req;
    this.res = res;
    this.response = false;//标记是否已经回复客户端信息
    return this;
};


Modular.prototype.display = _display;
Modular.prototype.cookie = _cookie;
Modular.prototype.session = _session;
Modular.prototype.send = _send;
Modular.prototype.render = _render;
Modular.prototype.param = _param;
Modular.prototype.files = _files.files;
Modular.prototype.path = _path;

Modular.prototype['200'] = _state['200'];
Modular.prototype['304'] = _state['304'];
Modular.prototype['404'] = _state['404'];
Modular.prototype['500'] = _state['500'];

Modular.fn = Modular.prototype;
Modular.fn.init.prototype = Modular.fn;

module.exports = Modular;
