// harbors/site_page/speech/Gruntfile.js

module.exports = function(grunt) {

    var transport = require('grunt-cmd-transport');
    var style = transport.style.init(grunt);
    var text = transport.text.init(grunt);
    var script = transport.script.init(grunt);

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),

        transport: {
            options:{
                idleading:'/script/compress/',
                alias:'<%= pkg.spm.alias%>',
                parsers : {
                    '.js' : [script.jsParser],
                    '.css' : [style.css2jsParser],
                    '.html' : [text.html2jsParser]
                }
            },
            build: {
                files : [
                    {
                        expand:true,
                        cwd:'script/src/',
                        src : ['**/*.js','**/*.css'],
                        dest : 'script/build/'
                    }
                ]
            }
        },
        concat: {
            options:{
                include:'relative'
            },
            merge: {
                files: [
                    {
                        expand: true,
                        cwd: 'script/build/',
                        src: ['**/*.js','!**/*-debug.js','!**/*-debug.css.js'],
                        dest: 'script/merge/'
                    }
                ]
            }
        },
        uglify: {
            compress: {
                files: [
                    {
                        expand: true,
                        cwd: 'script/merge/',
                        src: ['*/*.js','!mod/*.js'],
                        dest: 'script/compress/',
                        ext: '.js'
                    }
                ]
            }
        },
        clean : ['script/build','script/merge'],
        watch: {
            script:{
                files:['script/src/*/*.js','script/src/*/*css'],
                tasks:['transport','concat','uglify','clean']
            }
        },

        compass: {// compass
            src: {// Another target
                options: {
                    sassDir: 'style/sass',
                    cssDir: 'style/src',
                    outputStyle: 'expanded',
                    debugInfo: true
                },
                files: [
                    {
                        expand: true,
                        cwd: 'style/sass',
                        src: ['***/*//*.scss', '!*//*mod*//***/*//*.scss'],
                        dest: 'style/src',
                        ext: '.css'
                    }
                ]
            },
            dist: {// Target
                options: {// Target options
                    sassDir: 'style/sass',
                    cssDir: 'style/compress',
                    imageDir: 'image/src/',
                    environment: 'production',
                    debugInfo:false
                },
                files: [
                    {
                        expand: true,
                        cwd: 'style/sass',
                        src: ['**/*.scss', '!*/mod/**/*.scss'],
                        dest: 'style/compress',
                        ext: '.css'
                    }
                ]
            }
        }
    });

    grunt.loadNpmTasks('grunt-cmd-transport');
    grunt.loadNpmTasks('grunt-cmd-concat');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-watch');

    grunt.registerTask('js', function(){
        grunt.task.run('transport');
        grunt.task.run('concat');
        grunt.task.run('uglify');
        grunt.task.run('clean');
    });


    grunt.loadNpmTasks('grunt-contrib-compass');

    grunt.registerTask('css', ['compass']);
};
