// harbors/config/config.js
var path = require('path');
module.exports = {
    server:{
        dir             :   path.join(__dirname, '../../../', './node_modules/harbors/site_page/'),
        default_file    :   ['index.html', 'index_.js','index.php'],
        process         :   2,
        access_log_dir  :   path.join(__dirname, '../../../', './log/'),
        error_log_dir   :   path.join(__dirname, '../../../', './log/'),
        log_size        :   256,
        tmp_dir         :   path.join(__dirname, '../../../', './tmp/'),
        listen          :   [80, 3000, 3001],
        admin:{
            ip:["127.0.0.1","192.168.66.235"]
        },
        vhost       :   {
            on : true
        },
        cache       :   {
            on : false,
            expires : 86400000,
            cache_time : 1800000
        },
        socket:{
            on:true
        },
        zip         :   {
            on : true,
            zip_file : ['html', 'css', 'js']
        },
        rewrite     :   {
            on : false,
            name : ''
        },
        nodejs      :   {
            on : true,
            extname : '_.js',
            session : false,
            session_id : 'HsessID',
            session_survavil : 3600000,
            autoReload:true
        },
        php         :   {
            on : true,
            extname : '.php',
            fastcgi_host    :   '127.0.0.1',
            fastcgi_port    :   9000
        }
    },
    vhost:[]
};
