// harbors/config/config.js
var path = require('path');
module.exports = {
    server:{
        dir             :   path.join(__dirname, '../../../', './node_modules/harbors/site_page/'),
        default_file    :   ['index.html', 'index_.js','index.php'],
        process         :   1,
        access_log_dir  :   path.join(__dirname, '../../../', './log/'),
        error_log_dir   :   path.join(__dirname, '../../../', './log/'),
        log_size        :   256,
        tmp_dir         :   path.join(__dirname, '../../../', './tmp/'),
        listen          :   [3000,3001],
        support         :   [],
        vhost       :   {
            on : false
        },
        cache       :   {
            on : false,
            expires : 86400000,
            cache_time : 1800000
        },
        zip         :   {
            on : true,
            zip_file : ['html', 'css', 'js']
        },
        rewrite     :   {
            on : false,
            name : ''
        },
        nodejs      :   {
            on : true,
            extname : '_.js',
            cookie : false,
            session : false,
            session_id : 'HsessID',
            session_survavil : 3600000,
            post : false
        },
        php         :   {
            on : true,
            extname : '.php',
            fastcgi_host    :   '127.0.0.1',
            fastcgi_port    :   9000
        },
        module_extname  :   {
            nodejs  :   '_.js',
            php     :   '.php'
        }
    },
    vhost:[]
};
